ArduPlane/commands_logic.cpp

```diff
@@ -526,17 +526,33 @@ bool Plane::verify_takeoff()
             // course. This keeps wings level until we are ready to
             // rotate, and also allows us to cope with arbitrary
             // compass errors for auto takeoff
-            float takeoff_course = wrap_PI(radians(gps.ground_course_cd()*0.01f)) - steer_state.locked_course_err;
-            takeoff_course = wrap_PI(takeoff_course);
-            steer_state.hold_course_cd = wrap_360_cd(degrees(takeoff_course)*100);
+            const float ground_course_rad = wrap_PI(ahrs.groundspeed_vector().angle());
+            const float ground_course = wrap_360(degrees(ground_course_rad));
+            const float takeoff_course_rad = wrap_PI(ground_course_rad - steer_state.locked_course_err);
+
+            steer_state.hold_course_cd = wrap_360_cd(degrees(takeoff_course_rad)*100);
+
+            // when true, we will follow the recorded takeoff heading.
+            // when false, we xtrack to a virtual waypoint 10km ahead of us on this takeoff heading
+            // TODO: turn this hardcoded value into a param/option?
+            steer_state.locked_course = false;
+
+            next_WP_loc = current_loc;
+            next_WP_loc.offset_bearing(ground_course, 10000); // push it out 10km
+            set_next_WP(next_WP_loc);
+            auto_state.crosstrack = true;
+
             gcs().send_text(MAV_SEVERITY_INFO, "Holding course %d at %.1fm/s (%.1f)",
                               (int)steer_state.hold_course_cd,
                               (double)gps.ground_speed(),
                               (double)degrees(steer_state.locked_course_err));
         }
     }
 
-    if (steer_state.hold_course_cd != -1) {
+    if (!steer_state.locked_course && !prev_WP_loc.same_latlon_as(next_WP_loc)) {
```

Comment:
add `current_loc.alt > takeoff_loc.alt + 10m` and also don't push out the waypoint so far. Instead of a super-far hopefully unreachable target, it should be shorted (like 500m) recalculated (pushed out another 500m) if we get close

Reply:
This is just a thinking-outloud comment. Not necessary.